### PR TITLE
Jwt token [WIP]

### PR DESCRIPTION
**Pull request by agatatalita**

### CODE REVIEW

**Comment by mrkvon** [comment]
@agatatalita 
I just rebased the jwt-token branch with master, so jwt-token is up to date with master now.
`git pull origin jwt-token` won't work for you now.

You will need to do `git fetch origin jwt-token` and `git reset --hard origin/jwt-token` instead.
**This will overwrite your local jwt-token branch!**
If you have any commits locally since 13ac42afd5 you can create a new branch before overwriting `git branch [branchname]`.
If you have any uncommitted changes, you can `git stash` before overwriting and `git stash pop` after.
https://stackoverflow.com/a/8888015/4551707

File: test/users.username.js [hunk 1]
```diff
@@ -45,19 +59,19 @@ describe('/users/:username', function () {
     context('[user exists]', function () {
       it('[logged] should read user`s profile', async function () {
         const response = await agent
-          .get(`/users/${existentUser.username}`)
+          .get(`/users/${loggedUser.username}`)
```
Comment (mrkvon) [inline test/users.username.js]: why the change?

File: controllers/authorize.js [hunk 2]
```diff
@@ -1,26 +1,124 @@
...
+
+const contacts = require('./contacts');
+
+async function setAuthData(req, res, next) {
+  let token;
+  const user = {logged: false};
+  if (_.has(req, 'headers.authorization')){
+    token = req.headers.authorization;
+    req.auth = {};
+    const check = await tokenGetData(token);
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authorize.js]: Yes, we can do this here. So when check.valid is false, we will do return res.status(403).json({ errors: ['Not Authorized'] });

File: controllers/authorize.js [hunk 3]
```diff
@@ -1,26 +1,124 @@
...
+    }
+  }
   return res.status(403).json({ errors: ['Not Authorized'] });
   // TODO improve the error
-};
+}
+
 
 // Authorize only logged user who requests herself in /users/:username
-exports.onlyLoggedMe = function (req, res, next) {
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authorize.js]: ok

File: package.json [hunk 4]
```diff
@@ -7,6 +7,11 @@
     "start": "NODE_ENV=development node ./bin/www",
     "test": "NODE_ENV=test mocha ./test/",
     "test:watch": "NODE_ENV=test mocha ./test/ --watch",
+    "unit": "NODE_ENV=test mocha ./test/unit/",
+    "unit:watch": "NODE_ENV=test mocha ./test/unit/ --watch",
+    "test-all": "NODE_ENV=test mocha ./test/ --recursive",
+    "test-all:watch": "NODE_ENV=test mocha ./test/ --recursive --watch",
+    "nyc": "NODE_ENV=test nyc mocha ./test/ --recursive",
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply package.json]: unit tests are fast, aren't they? : )

File: package.json [hunk 5]
```diff
@@ -15,19 +20,23 @@
...
     "fs-extra": "^4.0.1",
     "handlebars": "^4.0.10",
     "he": "^1.1.1",
     "helmet": "^3.8.2",
     "identicon.js": "^2.1.0",
     "image-type": "^3.0.0",
     "jsonapi-serializer": "^3.4.1",
+    "jsonwebtoken": "^8.0.1",
     "lodash": "^4.16.4",
+    "lolex": "^2.1.2",
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply package.json]: no, I just thought the name is funny and I installed it.

File: package.json [hunk 6]
```diff
@@ -37,13 +46,18 @@
     "passport-oauth2": "^1.3.0",
     "prompt": "^1.0.0",
     "sanitize-html": "^1.14.1",
+    "samsam": "^1.2.1",
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply package.json]: as above

File: test/unit/auth/authenticate-jwt-token.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+      sinon = require('sinon');
+
+
+const models = require(path.resolve('./models'));
+const authController = require(path.resolve('./controllers/authenticate-token'));
+const jwtConfig = require(path.resolve('./config/secret/jwt-config'));
+
+
+
+describe.only('generateTokenBehavior(req)', function() {
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply test/unit/auth/authenticate-jwt-token.js]: I was/am experimenting with the best unit testing techniques for middlewares, and here I was separating behavior of generateToken middleware from actual middleware. Just here because I found it a little too confusing to continue with this practice in more places.

File: test/unit/router-tokenAuth.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+/* 'use strict';
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply test/unit/router-tokenAuth.js]: yes, and add router unit tests to it

File: controllers/authorize.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply controllers/authorize.js]: Yes.
Following JSON API [specification](http://jsonapi.org/format/#errors) and [examples](http://jsonapi.org/examples/#error-objects) the error body may look like this:

```
{
  errors: [{
    status: '403',
    title: 'Not Authorized'
  }]
}
```
Sometimes `detail: 'Expired'` would be useful for app. Is the "expired" information already available, or is it additional work?

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply package.json]: And the test coverage report is very interesting.

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply package.json]: really? :D
i believe sinon uses (or at least suggests) it for faking time. i read it in their documentation. :)

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply package.json]: omg :laughing:

File: test/unit/router-tokenAuth.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrkvon) [reply test/unit/router-tokenAuth.js]: how do you unit test express router? it's beyond my imagination. :)

File: controllers/authorize.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authorize.js]: I think it can easily be taken from jwt.verify error.

File: controllers/authorize.js [hunk 9]
```diff
@@ -1,26 +1,124 @@
...
+const jwtConfig = require(path.resolve('./config/secret/jwt-config'));
+
+const contacts = require('./contacts');
+
+async function setAuthData(req, res, next) {
+  let token;
+  const user = {logged: false};
+  if (_.has(req, 'headers.authorization')){
+    token = req.headers.authorization;
+    req.auth = {};
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authorize.js]: ok, you can edit it

File: controllers/authorize.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authorize.js]: I changed it but BDD tests for it are are missing

[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authorize.js]: You can change it. There will be two kind of headers: Basic and Token I think.

File: controllers/authenticate-token.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      models = require(path.resolve('./models'));
+
+async function generateTokenBehavior(data) {
+  let username, password, token, authenticated, verified, user;
+  const responseData = {};
+  // checks for authorisation headers
+  // and take username and password from header
+  if (_.has(data, 'headers.authorization')) {
+    try{
+      username = auth(data).name;
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authenticate-token.js]: changed

File: controllers/authenticate-token.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    // TODO kind of error to return
+    if(authenticated === undefined || verified === undefined) {
+      responseData.status = 401;
+      responseData.message = 'Invalid user object';
+      responseData.data = {};
+      return responseData;
+    }
+
+    if(!authenticated) {
+      responseData.status = 403;
```
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/authenticate-token.js]: changed

File: controllers/authenticate-token.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    responseData.data = {};
+    return responseData;
+  }
+
+  try{
+    // checks if user exists
+    const userExists = await models.user.exists(username);
+    if (!userExists || username === undefined || password === undefined) {
+      responseData.status = 401;
+      responseData.message = 'User doesn\'t exist or lack of username or password';
```
Comment (mrkvon) [inline controllers/authenticate-token.js]: For security reasons all 401 messages should be the same, not to reveal any information to an attacker. https://www.owasp.org/index.php/Authentication_Cheat_Sheet#Correct_Response_Example

**Comment by mrkvon** [comment]
:clap: for all the great work @agatatalita
